File: Week_04/java/src/_130_SurroundedRegions.java
/**
 * 第一遍：2020/08/11周二 ✅
 * 第二遍：2020/08/10周二
 * 第三遍：2020/07/18周六
 * 第三遍：2020/07/08周四
 * 菜。。
 */
class _130_SurroundedRegions {
    public void solve(char[][] board) {
        if (board == null) return;
        int m = board.length;
        if (m < 3) return;
        int n = board[0].length;
        if (n < 3) return;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 'O' && (i == 0 || i == m - 1 || j == 0 || j == n - 1)) {
                    dfs(board, i, j);
                }
            }
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == '#') {
                    board[i][j] = 'O';
                } else if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }
            }
        }
    }

    int[][] dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    void dfs(char[][] board, int i, int j) {
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || board[i][j] == 'X' || board[i][j] == '#') {
            return;
        }
        board[i][j] = '#';
        for (int[] dir : dirs) {
            dfs(board, i + dir[0], j + dir[1]);
        }
    }
}

File: Week_04/kotlin/src/_130_SurroundedRegions.kt
fun solve(board: Array<CharArray>): Unit {
    val dirs = arrayOf(intArrayOf(1, 0), intArrayOf(-1, 0), intArrayOf(0, 1), intArrayOf(0, -1))
    fun dfs(board: Array<CharArray>, i: Int, j: Int) {
        if (i !in board.indices || j !in board[0].indices || board[i][j] == 'X' || board[i][j] == '#') {
            return
        }

        board[i][j] = '#'
        for (dir in dirs) {
            dfs(board, i + dir[0], j + dir[1])
        }
    }

    val m = board.size.takeIf { it >= 3 } ?: return
    val n = board[0].size.takeIf { it >= 3 } ?: return

    for (i in intArrayOf(0, m - 1)) {
        for (j in 0 until n) {
            if (board[i][j] == 'O') {
                dfs(board, i, j)
            }
        }
    }

    for (j in intArrayOf(0, n - 1)) {
        for (i in 1..m - 2) {
            if (board[i][j] == 'O') {
                dfs(board, i, j)
            }
        }
    }

    for (i in 0 until m) {
        for (j in 0 until n) {
            when (board[i][j]) {
                'O' -> {
                    board[i][j] = 'X'
                }
                '#' -> {
                    board[i][j] = 'O'
                }
            }
        }
    }
}